lib/modules/manager/maven-wrapper/artifacts.ts

```diff
@@ -0,0 +1,202 @@
+import type {
+  UpdateArtifact,
+  UpdateArtifactsConfig,
+  UpdateArtifactsResult,
+} from '../types';
+
+async function addIfUpdated(
+  status: StatusResult,
+  fileProjectPath: string
+): Promise<UpdateArtifactsResult | null> {
+  if (status.modified.includes(fileProjectPath)) {
+    return {
+      file: {
+        type: 'addition',
+        path: fileProjectPath,
+        contents: await readLocalFile(fileProjectPath),
+      },
+    };
+  }
+  return null;
+}
+
+export async function updateArtifacts({
+  packageFileName,
+  newPackageFileContent,
+  updatedDeps,
+  config,
+}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
+  try {
+    logger.debug({ updatedDeps }, 'maven-wrapper.updateArtifacts()');
+
+    if (
+      !updatedDeps.some(
+        (dep) => dep.depName === 'org.apache.maven.wrapper:maven-wrapper'
+      )
+    ) {
+      logger.info(
+        'Maven wrapper version not updated - skipping Artifacts update'
+      );
+      return null;
+    }
+
+    const cmd = await createWrapperCommand(packageFileName);
+
+    if (!cmd) {
+      logger.info('No mvnw found - skipping Artifacts update');
+      return null;
+    }
+    await executeWrapperCommand(cmd, config);
+
+    const status = await getRepoStatus();
+    const artifactFileNames = [
+      '.mvn/wrapper/maven-wrapper.properties',
+      '.mvn/wrapper/maven-wrapper.jar',
+      '.mvn/wrapper/MavenWrapperDownloader.java',
+      'mvnw',
+      'mvnw.cmd',
+    ].map(
+      (filename) =>
+        packageFileName.replace('.mvn/wrapper/maven-wrapper.properties', '') +
+        filename
+    );
+    const updateArtifactsResult = (
+      await getUpdatedArtifacts(status, artifactFileNames)
+    ).filter(is.truthy);
+
+    logger.debug(
+      { files: updateArtifactsResult.map((r) => r.file?.path) },
+      `Returning updated maven-wrapper files`
+    );
+    return updateArtifactsResult;
+  } catch (err) {
+    logger.debug({ err }, 'Error setting new Maven Wrapper release value');
+    return [
+      {
+        artifactError: {
+          lockFile: packageFileName,
+          stderr: err.message,
+        },
+      },
+    ];
+  }
+}
+
+async function getUpdatedArtifacts(
+  status: StatusResult,
+  artifactFileNames: string[]
+): Promise<UpdateArtifactsResult[]> {
+  const updatedResults: (UpdateArtifactsResult | null)[] = [];
+  for (const artifactFileName of artifactFileNames) {
+    updatedResults.push(await addIfUpdated(status, artifactFileName));
+  }
+  return updatedResults.filter(is.truthy);
+}
+
+/**
+ * Find compatible java version for maven.
+ * see https://maven.apache.org/developers/compatibility-plan.html
+ * @param mavenWrapperVersion current maven version
+ * @returns A Java semver range
+ */
+export function getJavaConstraint(
+  mavenWrapperVersion: string | null | undefined
+): string | null {
+  const major = mavenWrapperVersion
+    ? mavenVersioning.getMajor(mavenWrapperVersion)
+    : null;
+
+  if (major && major >= 3) {
```

Comment:
are you sure maven 3.0.0 already supports java 17? It was released at `2010-10-08` so i'm pretty sure it's not. Please validate with real repo tests

Reply:
The support should mostly goes the other way right?  Java should be backwards compatible enough to run maven 3.0.0, but will test it in practice.  Good feedback, thanks.

Reply:
did you validated this?

Reply:
Yes, we tested it both locally and in repositories in github & gitlab.

There is no problem with java backwards compatibility.

However, there are some newer versions of the wrapper that do not work with older versions of maven, but we don't think that's a problem since you can always manually upgrade both to fix this. Is that OK?